### PR TITLE
Lance Nelson

### PR DESCRIPTION
**Pull request by LancePants97**
Spent about an hour trying to figure out what was wrong with my code and wondering why my many to many relationship was broken, only to find that a couple words were plural that shouldn't be...anyways I got user stories 1 and 2 completed and feel like I could have had a good shot at user story 3 if I had not wasted so much time!!

### CODE REVIEW

File: spec/features/dishes/show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+# And I see the chef's name.
+      it "I see the dish's name, description, ingredients, calorie count, and chef's name" do
+        visit "/dishes/#{@dish.id}"
+
+        expect(page).to have_content("Dish: #{@dish.name}")
+        expect(page).to have_content(@dish.description)
+        expect(page).to have_content(@chef.name)
+        expect(page).to have_content(@ingredient.name)
+        expect(page).to have_content(@ingredient2.name)
+        expect(page).to have_content(@ingredient3.name)
```
Comment (thisfncodeio) [inline spec/features/dishes/show_spec.rb:38]: Is there any reason why you added anchor text to just the dish name on line 33 and not the rest? I would have really liked to see consistency across all of your assertions.

File: spec/features/dishes/show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        total_calories = (@ingredient.calories + @ingredient2.calories + @ingredient3.calories) 
+        expect(page).to have_content(total_calories)
```
Comment (thisfncodeio) [inline spec/features/dishes/show_spec.rb:41]: `expect(page).to have_content(800)` ... this is your assertion on line 41. I feel like we've been calling out this type of assertion before. This is not a good assertion. Numbers have the possibility to produce false positives, and we want to make sure that we are writing our assertions to guard against this, with anchor text or labels, or `within` blocks.

File: spec/features/dishes/show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      it "allows me to add ingredients by ID" do
+        visit "/dishes/#{@dish.id}"
+      
+        fill_in :ingredient_id, with: @ingredient4.id
+        
+        click_button :submit
+
+        expect(current_path).to eq("/dishes/#{@dish.id}")
+        expect(page).to have_content(@ingredient4.name)
+      end
```
Comment (thisfncodeio) [inline spec/features/dishes/show_spec.rb:59]: While this test is ok and covers the user story, one way to improve it is to write and assertions that proves that Ingredient 4 wasn't already on the page.

```ruby
it "allows me to add ingredients by ID" do
  visit "/dishes/#{@dish.id}"
  expect(page).to_not have_content(@ingredient4.name)  

  fill_in :ingredient_id, with: @ingredient4.id

  click_button :submit

  expect(current_path).to eq("/dishes/#{@dish.id}")
  expect(page).to have_content(@ingredient4.name)
end
```

File: config/routes.rb [hunk 4]
```diff
@@ -3,4 +3,6 @@
 
   # Defines the root path route ("/")
   # root "articles#index"
+  get "/dishes/:id", to: "dishes#show"
+  post "/dishes/:id/ingredients", to: "ingredient_dishes#create"
 end
```
Comment (thisfncodeio) [inline config/routes.rb:8]: Great restful routing and sending the request to the proper controller for joining an ingredient to a dish.

File: app/models/dish.rb [hunk 5]
```diff
@@ -1,4 +1,6 @@
 class Dish < ApplicationRecord
   validates_presence_of :name, :description
   belongs_to :chef
+  has_many :ingredient_dishes
+  has_many :ingredients, through: :ingredient_dishes
 end
```
Comment (thisfncodeio) [inline app/models/dish.rb:6]: I do not see any additional model tests for these added assertions.

File: app/models/ingredient.rb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+class Ingredient < ApplicationRecord
+  has_many :ingredient_dishes
+  has_many :dishes, through: :ingredient_dishes
+end
```
Comment (thisfncodeio) [inline app/models/ingredient.rb:4]: I do not see any additional model tests for these added assertions.

File: app/models/ingredient_dish.rb [hunk 7]
```diff
@@ -0,0 +1,4 @@
+class IngredientDish < ApplicationRecord
+  belongs_to :dish
+  belongs_to :ingredient
+end
```
Comment (thisfncodeio) [inline app/models/ingredient_dish.rb:4]: I do not see any additional model tests for these added assertions.

File: app/controllers/ingredient_dishes_controller.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+class IngredientDishesController < ApplicationController
+  def create
+    dish = Dish.find(params[:id])
+    ingredient = Ingredient.find(params[:ingredient_id])
+    IngredientDish.create(ingredient: ingredient, dish: dish)
+    redirect_to "/dishes/#{dish.id}"
+  end
+end
```
Comment (thisfncodeio) [inline app/controllers/ingredient_dishes_controller.rb:8]: Great job with this, this is 100% the correct action and controller that should be handling this request.

File: app/views/dishes/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+<h1>Dish: <%= @dish.name %></h1>
+<h2><%= @dish.description %></h2>
+<h3>Chef: </h3><%= @dish.chef.name %>
+<% total_calories = 0 %>
+<h3>Ingredients: </h3><% @dish.ingredients.each do |ingredient| %>
+    <%= ingredient.name %><br>
+    <% total_calories += ingredient.calories %>
+<% end %>
+<h3>Total Calories: </h3><%= total_calories %>
```
Comment (thisfncodeio) [inline app/views/dishes/show.html.erb:9]: The way you are getting the total calories is incorrect. You are using Ruby to process database data over using pure Active Record. This code should be using all AR and be abstracted to a model method.

File: app/views/dishes/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+<h3>Add More of an Existing Ingredient</h3>
+<%= form_with url: "/dishes/#{@dish.id}/ingredients", method: :post, data: { turbo: false } do |form| %>
+  <%= form.label :ingredient_id %>
+  <%= form.text_field :ingredient_id %>
+  <%= form.submit :submit %>
+<% end %>
```
Comment (thisfncodeio) [inline app/views/dishes/show.html.erb:16]: This form is perfect.

**Review by thisfncodeio — COMMENTED** [review]
It looks like most of your implementation for US 1 was correct, except for the way you are getting the total calories for all the ingredients in the dish. US 2 was implemented properly.

I would encourage you to work on improving the rigidity of your tests. This can easily be done using anchor/label text or `within` blocks. Don't forget your model tests as well... Any association you add should be tested in order for your model to be considered 100% tested.

**Comment by abdulredd** [comment]
Just realized I was on my personal account when I left feedback--oops.

**Pull request closed without merging**